File: resources/common/org/webpki/saturn/common/TransactionRequest.java
/*
 *  Copyright 2006-2016 WebPKI.org (http://webpki.org).
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */
package org.webpki.saturn.common;

import java.io.IOException;

import java.math.BigDecimal;

import java.security.GeneralSecurityException;

import java.util.GregorianCalendar;

import java.util.Vector;

import org.webpki.json.JSONObjectReader;
import org.webpki.json.JSONObjectWriter;
import org.webpki.json.JSONParser;
import org.webpki.json.JSONSignatureDecoder;
import org.webpki.json.JSONX509Verifier;

import org.webpki.json.encryption.DecryptionKeyHolder;

public class TransactionRequest implements BaseProperties {
    
    public TransactionRequest(JSONObjectReader rd, Boolean cardNetwork) throws IOException {
        root = Messages.TRANSACTION_REQUEST.parseBaseMessage(rd);
        authorizationResponse = new AuthorizationResponse(Messages.AUTHORIZATION_RESPONSE.getEmbeddedMessage(rd));
        recepientUrl = rd.getString(RECEPIENT_URL_JSON);
        actualAmount = rd.getBigDecimal(AMOUNT_JSON,
                                        authorizationResponse
                                            .authorizationRequest.paymentRequest.currency.decimals);
        referenceId = rd.getString(REFERENCE_ID_JSON);
        timeStamp = rd.getDateTime(TIME_STAMP_JSON);
        software = new Software(rd);
        signatureDecoder = rd.getSignature(new JSONSignatureDecoder.Options());
        if (cardNetwork != null &&
            authorizationResponse.authorizationRequest.payerAccountType.isCardPayment() ^ cardNetwork) {
            throw new IOException("Incompatible payment method: " + 
                authorizationResponse.authorizationRequest.payerAccountType.getTypeUri());
        }
        rd.checkForUnread();
    }

    Software software;

    JSONObjectReader root;

    String recepientUrl;
    public String getRecepientUrl() {
        return recepientUrl;
    }

    GregorianCalendar timeStamp;
    public GregorianCalendar getTimeStamp() {
        return timeStamp;
    }

    BigDecimal actualAmount;
    public BigDecimal getAmount() {
        return actualAmount;
    }

    String referenceId;
    public String getReferenceId() {
        return referenceId;
    }

    public boolean getTestMode() {
        return authorizationResponse.authorizationRequest.testMode;
    }

    public Payee getPayee() {
        return authorizationResponse.authorizationRequest.paymentRequest.payee;
    }

    public PaymentRequest getPaymentRequest() {
        return authorizationResponse.authorizationRequest.paymentRequest;
    }

    JSONSignatureDecoder signatureDecoder;
    public JSONSignatureDecoder getSignatureDecoder() {
        return signatureDecoder;
    }

    AuthorizationResponse authorizationResponse;
    public AuthorizationResponse getAuthorizationResponse() {
        return authorizationResponse;
    }

    public static JSONObjectWriter encode(AuthorizationResponse authorizationResponse,
                                          String recepientUrl,
                                          BigDecimal actualAmount,
                                          String referenceId,
                                          ServerAsymKeySigner signer) throws IOException {
        return Messages.TRANSACTION_REQUEST.createBaseMessage()
            .setObject(Messages.AUTHORIZATION_RESPONSE.lowerCamelCase(), authorizationResponse.root)
            .setString(RECEPIENT_URL_JSON, recepientUrl)
            .setBigDecimal(AMOUNT_JSON,
                           actualAmount,
                           authorizationResponse.authorizationRequest.paymentRequest.currency.decimals)
            .setString(REFERENCE_ID_JSON, referenceId)
            .setDateTime(TIME_STAMP_JSON, new GregorianCalendar(), true)
            .setObject(SOFTWARE_JSON, Software.encode(PaymentRequest.SOFTWARE_NAME, 
                                                      PaymentRequest.SOFTWARE_VERSION))
            .setSignature(signer);
    }

    public ProtectedAccountData getProtectedAccountData(Vector<DecryptionKeyHolder> decryptionKeys)
    throws IOException, GeneralSecurityException {
        return new ProtectedAccountData(JSONParser.parse(authorizationResponse
                                                             .encryptedAccountData
                                                                 .getDecryptedData(decryptionKeys)),
                                        authorizationResponse.authorizationRequest.payerAccountType);
    }

    public void verifyPayerBank(JSONX509Verifier paymentRoot) throws IOException {
        authorizationResponse.signatureDecoder.verify(paymentRoot);
    }
}

File: nodejs-common/TransactionRequest.js
/*
 *  Copyright 2015-2020 WebPKI.org (http://webpki.org).
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

'use strict';

// Saturn "TransactionRequest" object

const JsonUtil  = require('webpki.org').JsonUtil;
const ByteArray = require('webpki.org').ByteArray;

const BaseProperties        = require('./BaseProperties');
const Messages              = require('./Messages');
const Software              = require('./Software');
const AuthorizationResponse = require('./AuthorizationResponse');
const AuthorizationRequest  = require('./AuthorizationRequest');
const AccountDataDecoder    = require('./AccountDataDecoder');

function TransactionRequest(rd) {
  this.root = Messages.parseBaseMessage(Messages.TRANSACTION_REQUEST, rd);
  this.authorizationResponse = new AuthorizationResponse(Messages.getEmbeddedMessage(Messages.AUTHORIZATION_RESPONSE, rd));
  this.recipientUrl = rd.getString(BaseProperties.RECIPIENT_URL_JSON);
  this.actualAmount = rd.getBigDecimal(BaseProperties.AMOUNT_JSON,
                                       this.authorizationResponse.authorizationRequest.paymentRequest.currency.decimals);
  this.referenceId = rd.getString(BaseProperties.REFERENCE_ID_JSON);
  this.timeStamp = rd.getDateTime(BaseProperties.TIME_STAMP_JSON);
  this.software = new Software(rd);
  this.publicKey = rd.getSignature(BaseProperties.REQUEST_SIGNATURE_JSON).getPublicKey();
  rd.checkForUnread();
}

TransactionRequest.prototype.getTimeStamp = function() {
  return this.timeStamp;
};

TransactionRequest.prototype.getReferenceId = function() {
  return this.referenceId;
};

TransactionRequest.prototype.getAmount = function() {
  return this.actualAmount;
};

TransactionRequest.prototype.getTestMode = function() {
  return this.authorizationResponse.authorizationRequest.testMode;
};

TransactionRequest.prototype.getPayeeAuthorityUrl = function() {
  return this.authorizationResponse.authorizationRequest.payeeAuthorityUrl;
};

TransactionRequest.prototype.getPublicKey = function() {
  return this.publicKey;
};

TransactionRequest.prototype.getAuthorizationPublicKey = function() {
  return this.authorizationResponse.authorizationRequest.publicKey;
};

TransactionRequest.prototype.getAuthorizationResponse = function() {
  return this.authorizationResponse;
};

TransactionRequest.prototype.getPaymentRequest = function() {
  return this.authorizationResponse.authorizationRequest.paymentRequest;
};

TransactionRequest.prototype.verifyPayerProvider = function(paymentRoot) {
  this.authorizationResponse.signatureDecoder.verifyTrust(paymentRoot);
};

TransactionRequest.prototype.getProtectedCardData = function(decryptionKeys) {
  let cardData = new AccountDataDecoder(JsonUtil.ObjectReader.parse(
                               this.authorizationResponse.encryptedAccountData
                                          .getDecryptedData(decryptionKeys)));
  cardData.checkAccountTypes(AccountDataDecoder.SUPERCARD_ACCOUNT);
  return cardData;
};

module.exports = TransactionRequest;
